File: Models/Education.cs

using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace JobFairPortal.Models
{
    public class Education
    {
        [Key]
        public int EducationId { get; set; }

        [Required]
        [ForeignKey("Student")]
        public int StudentId { get; set; }

        [Required]
        [MaxLength(150)]
        public string InstitutionName { get; set; } = null!;

        [Required]
        [MaxLength(100)]
        public string Degree { get; set; } = null!; 

        [MaxLength(100)]
        public string? FieldOfStudy { get; set; }  

        public DateTime? StartDate { get; set; }

        public DateTime? EndDate { get; set; }

        public bool IsCurrent { get; set; } = false;

        [Range(0, 4.0)]
        public double? CGPA { get; set; }

        [MaxLength(200)]
        public string? Location { get; set; }

        public Student Student { get; set; } = null!;
    }
}

File: lib/model/education.dart
import 'package:student_job_fair_portal/model/student.dart';

class Education {
  final int educationId;
  final int studentId;
  final String institutionName;
  final String degree;
  final String? fieldOfStudy;
  final DateTime? startDate;
  final DateTime? endDate;
  final bool isCurrent;
  final double? cgpa;
  final String? location;

  Education({
    required this.educationId,
    required this.studentId,
    required this.institutionName,
    required this.degree,
    this.fieldOfStudy,
    this.startDate,
    this.endDate,
    required this.isCurrent,
    this.cgpa,
    this.location,
  });

  factory Education.fromJson(Map<String, dynamic> json) {
    return Education(
      // 🔹 FIX: Use '?? 0' to handle missing studentId without crashing
      educationId: json['educationId'] ?? 0,
      studentId: json['studentId'] ?? 0,

      institutionName: json['institutionName'] ?? '',
      degree: json['degree'] ?? '',
      fieldOfStudy: json['fieldOfStudy'],
      startDate: Student.safeParseDate(json['startDate']),
      endDate: Student.safeParseDate(json['endDate']),
      isCurrent: json['isCurrent'] ?? false,
      cgpa: (json['cgpa'] as num?)?.toDouble(),
      location: json['location'],
    );
  }

  // 🔹 Add copyWith to allow the Provider to fix the ID later
  Education copyWith({
    int? educationId,
    int? studentId,
    String? institutionName,
    String? degree,
    String? fieldOfStudy,
    DateTime? startDate,
    DateTime? endDate,
    bool? isCurrent,
    double? cgpa,
    String? location,
  }) {
    return Education(
      educationId: educationId ?? this.educationId,
      studentId: studentId ?? this.studentId,
      institutionName: institutionName ?? this.institutionName,
      degree: degree ?? this.degree,
      fieldOfStudy: fieldOfStudy ?? this.fieldOfStudy,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      isCurrent: isCurrent ?? this.isCurrent,
      cgpa: cgpa ?? this.cgpa,
      location: location ?? this.location,
    );
  }
}
